### PR TITLE
CL_USE_HOST_MEM_PTR with zero copy for edge

### PR DESCRIPTION
**Pull request by Himanshu-xilinx**
* Api to alloc/free mem aligned and continues host buffer for edge.
* Api to alloc/free mem aligned host buffer for pcie.
* hal api xclAllocHostPtr xclFreeHostPtr
* opencl api clAllocHostPtr clFreeHostPtr

### CODE REVIEW

**Comment by larry9523** [comment]
A couple of high level questions
1) I thought the problem we are trying to solve is, user application has a pre allocated buffer (not allocated by XRT interfaces), which might be not physically contiguous and 4K aligned, and they try to use that buffer by CL_USE_HOST_MEM_PTR flag. How would this new API can help with that?
2) Anything new that this API buy us if applications are not using xclAllocBO, xclMapBO and xclGetBOProperties?

File: src/runtime_src/core/edge/user/shim.cpp [hunk 1]
```diff
@@ -338,6 +350,42 @@ void *ZYNQShim::xclMapBO(unsigned int boHandle, bool write)
...
+{
+  std::cout << "ZYNQShim::xclFreeHostPtr called" << std::endl;
+  std::lock_guard<std::mutex> l(mBOMapLock);
+  auto it = mBoMap.find(reinterpret_cast<uint64_t>(ptr));
+  if (it == mBoMap.end()) {
+    std::cout  << "xclFreeHostPtr Bad Host PTR" << std::endl;
+    return;
+  }
+  munmap(ptr,it->second.size);  
+  xclFreeBO(it->second.boHandle);
```
Comment (stsoe) [inline src/runtime_src/core/edge/user/shim.cpp]: Wouldn't it be better to have BOData constructor and destructor manage xclAllocBO and xclFreeBO

File: src/runtime_src/core/include/xrt_helper.hpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+#include <vector>
+
+namespace xrt_util {
+
+class helper {
+    static helper* mInstance;
+	std::vector<xclDeviceHandle> mHandleList;
+	helper() {
+		xclDeviceInfo2 deviceInfo;
+		int ndevice = xclProbe();;
```
Comment (stsoe) [inline src/runtime_src/core/include/xrt_helper.hpp]: double ;;

File: src/runtime_src/core/include/xrt_helper.hpp [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    		if(ndevice  == 0) {
```
Comment (stsoe) [inline src/runtime_src/core/include/xrt_helper.hpp]: No curly braces around single statement.

File: src/runtime_src/core/include/xrt_helper.hpp [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+			throw std::runtime_error("No device found");
+		}
+
+		for(int i=0;i<ndevice; i++) {
+			xclDeviceHandle handle;
+			handle = xclOpen(i, NULL, XCL_INFO);
```
Comment (stsoe) [inline src/runtime_src/core/include/xrt_helper.hpp]: nullptr in C++
also why three lines of code?   just write ...push_back(xclOpen(...));

File: src/runtime_src/core/include/xrt_helper.hpp [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+			mHandleList.push_back(handle);
+		}
+	}	
+public:
+    static helper* getinstance(unsigned deviceIndex=0) {
+		if(mInstance == nullptr) {
+			mInstance = new helper();
```
Comment (stsoe) [inline src/runtime_src/core/include/xrt_helper.hpp]: Who will delete this?
Just use a static variable and return address of that.

File: src/runtime_src/core/include/xrt_helper.hpp [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+	~helper() {
+		for (auto it : mHandleList) {
+			xclClose(it);
+		}
+	}
+
+};
+
+helper* helper::mInstance = nullptr;
```
Comment (stsoe) [inline src/runtime_src/core/include/xrt_helper.hpp]: No needed.

File: src/runtime_src/core/edge/drm/zocl/zocl_bo.c [hunk 7]
```diff
@@ -427,16 +428,19 @@ zocl_userptr_bo_ioctl(struct drm_device *dev, void *data, struct drm_file *filp)
 	}
 
 	bo->cma_base.paddr = sg_dma_address((bo->cma_base.sgt)->sgl);
+*/
```
Comment (mamin506) [inline src/runtime_src/core/edge/drm/zocl/zocl_bo.c]: Why do we remove this check? What if user use a wrong pointer at the call? We would lost protect.

File: src/runtime_src/core/edge/drm/zocl/zocl_bo.c [hunk 8]
```diff
@@ -427,16 +428,19 @@ zocl_userptr_bo_ioctl(struct drm_device *dev, void *data, struct drm_file *filp)
...
+/*
 	if (sg_count != 1) {
 		DRM_ERROR("User buffer is not physical contiguous\n");
 		ret = -EINVAL;
 		goto out0;
 	}
 
 	bo->cma_base.vaddr = (void *)(uintptr_t)args->addr;
-
+*/
```
Comment (mamin506) [inline src/runtime_src/core/edge/drm/zocl/zocl_bo.c]: see above.

**Comment by sonals** [comment]
As @mamin506 and @larry9523 indicated let's take this offline and discuss how it solves the host ptr needs from specific applications.

**Comment by maxzhen** [comment]
@mamin506 will implement a work around for now.

**Pull request closed without merging**